helpers/prism.js

```diff
@@ -0,0 +1,432 @@
+};
+
+const languageDependencies = {
+	arduino: [ 'cpp' ],
+	cpp: [ 'c' ],
+	racket: [ 'scheme' ]
+};
+
+const languagesLoaded = {
+	clike: Promise.resolve(),
+	css: Promise.resolve(),
+	javascript: Promise.resolve(),
+	markup: Promise.resolve(),
+	plain: Promise.resolve()
+};
+
+const loadLanguage = async key => {
+	if (languagesLoaded[key]) return languagesLoaded[key];
+
+	// Prism languages can extend other anguages and must be loaded in order
+
+	// eslint-disable-next-line no-async-promise-executor
+	languagesLoaded[key] = new Promise(async resolve => {
+		if (languageDependencies[key]) {
+			await Promise.all(languageDependencies[key].map(dependencyKey => loadLanguage(dependencyKey)));
+		}
+
+		const script = document.createElement('script');
+		script.async = 'async';
+		script.onload = resolve;
+		script.src = `${prismLocation}/components/prism-${key}.min.js`;
+		document.head.appendChild(script);
+	});
+
+	return languagesLoaded[key];
+};
+
+const pluginsLoaded = {};
+
+const loadPlugin = async plugin => {
+	if (pluginsLoaded[plugin]) return pluginsLoaded[plugin];
+
+	pluginsLoaded[plugin] = new Promise(resolve => {
+		const script = document.createElement('script');
+		script.async = 'async';
+		script.onload = resolve;
+		script.src = `${prismLocation}/plugins/${plugin}/prism-${plugin}.min.js`;
+		document.head.appendChild(script);
+	});
+
+	return pluginsLoaded[plugin];
+};
+
+let prismLoaded;
+
+const loadPrism = () => {
+	if (prismLoaded) return prismLoaded;
+
+	prismLoaded = Promise.all([
+		new Promise(resolve => {
+			const script = document.createElement('script');
+			script.async = 'async';
+			script.onload = resolve;
+			script.src = `${prismLocation}/prism.js`;
+			document.head.appendChild(script);
+		}),
+		new Promise(resolve => {
+			const style = document.createElement('style');
+			style.textContent = codeStyles.cssText;
+			style.onload = resolve;
+			document.head.appendChild(style);
+		})
+	]);
+
+	return prismLoaded;
+};
+
+const getCodeElement = elem => {
+	if (!elem) return;
+	if (elem.tagName === 'CODE') return elem;
+	if (elem.tagName !== 'PRE') return;
+	return elem.querySelector('code');
+};
+
+export async function formatCodeElement(elem) {
+	const code = getCodeElement(elem);
+
+	if (code.className.indexOf('language-') === -1) return;
+
+	const languageInfo = getLanguageInfo(code);
+	const lineNumbers = elem.classList.contains('line-numbers') || code.classList.contains('line-numbers');
+
+	await loadPrism(); // must be loaded before loading plugins or languages
+	await Promise.all([
+		loadLanguage(languageInfo.key),
+		lineNumbers ? loadPlugin('line-numbers') : null
+	]);
+
+	if (!elem.dataset.language && languageInfo.key !== 'plain') elem.dataset.language = languageInfo.desc;
+	Prism.highlightElement(code);
+
+	return elem;
+}
+
+export class HtmlBlockCodeRenderer {
+
+	get canRenderInline() {
+		return true;
+	}
+
+	async render(elem) {
+		const codeElements = [...elem.querySelectorAll('.d2l-code')];
```

Comment:
We specifically only want to format elements that have `d2l-code` class.  Authors could have used `pre` or `code` elements for other reasons and we don't want to format those elements.